### PR TITLE
Table widget: default selected row, 0th

### LINKED ISSUE #5892: [Feature] Table widget: default selected row, 0th
## Summary

Set the table widget `default selected row` to 0 by default such that the first row of table is selected

## Motivation

Often the user is using `selectedRow` (which is not set), gets and error in the API response. We want to make it easy for the user to operate.

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Table widget: default selected row, 0th

Fixes #5892

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ]  I have commented my code, particularly in hard-to-understand areas
- [ ]  I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- []  I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feature/default-selected-row-zero 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.1 **(0.01)** | 37.01 **(0.02)** | 33.8 **(0.01)** | 55.7 **(0.01)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 74.55 **(0.52)** | 66.47 **(0.27)** | 69.89 **(0.66)** | 73.57 **(0.47)**</details>

### CODE REVIEW

**Review by somangshu — APPROVED** [review]
LGTM, @vivekverma2312 lets check if this caused any problem in old widgets, We will need to write migrations accordingly.

**Comment by vivekverma2312** [comment]
@techbhavin Couple of observations:
1. Default selected row property is still showing as blank, if default value is 0 then we should also mention it in the property
Screenshot 2021-08-02 at 12 25 48 PM


2. By default row is not selected for old tables.

**Comment by techbhavin** [comment]
@somangshu  @vivekverma2312  
> Point 1

There is an issue when propertyPane INPUT_TEXT type control has validation on Number. For default props and prop change from external source ( i.e. a btn click ) and new value of prop is "0" it is not reflected in propertyPane input field but new value of prop does reach out to component and it is updated as expected.

> Point 2 
we need to add migrations for this. let me add it

**Comment by techbhavin** [comment]
> By default row is not selected for old tables.

@somangshu  @vivekverma2312   migration test was added so please test point 2.

**Review by somangshu — APPROVED** [review]
Approving for the tests to run.

@marks0351 can you please verify the migrations?

**Review by marks0351 — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@techbhavin Migration issue is fixed and working fine. Issue 1 still exists.

**Comment by marks0351** [comment]
@techbhavin defaultSelectedRowValidation is the validation that runs to evaluate. so this doesn't work because by default the evaluated value will be returned as 0 and a check in the input text control to check value truthness fails. so we need to return out strings instead of numbers in defaultSelectedRowValidation. this doesnot happen when the user types the input because it is always handled as strings. let me know if you want to discuss this.

File: app/client/src/mockResponses/WidgetConfigResponse.tsx [hunk 1]
```diff
@@ -169,6 +169,7 @@ const WidgetConfigResponse: WidgetConfigReducerState = {
     TABLE_WIDGET: {
       rows: 7 * GRID_DENSITY_MIGRATION_V1,
       columns: 9 * GRID_DENSITY_MIGRATION_V1,
+      defaultSelectedRow: "0" as any,
```
Comment (marks0351) [inline app/client/src/mockResponses/WidgetConfigResponse.tsx:172]: can u pls change the interface instead, wud help catch errors in other areas of code flow.

**Review by marks0351 — CHANGES_REQUESTED** [review]
Please do add cypress or jest test cases as well.

**Review by marks0351 — APPROVED** [review]

**Comment by somangshu** [comment]
Upon testing the deploy preview, I see that the old tables do not have the change. I think we should write migration for this as well if not already done. @marks0351 what is your thought?

The feature otherwise works well for the new tables and the above mentioned issues from the QA are fixed, Once we discuss the aforementioned point we can take this ahead

File: app/client/src/utils/WidgetPropsUtils.tsx [hunk 2]
```diff
@@ -806,12 +806,31 @@ const transformDSL = (currentDSL: ContainerWidgetProps<WidgetProps>) => {
...
   }
 
   return currentDSL;
 };
 
+export const migrateTableDefaultSelectedRow = (
+  currentDSL: ContainerWidgetProps<WidgetProps>,
+) => {
+  if (currentDSL.type === WidgetTypes.TABLE_WIDGET) {
+    if (!currentDSL.defaultSelectedRow) currentDSL.defaultSelectedRow = 0;
```
Comment (marks0351) [inline app/client/src/utils/WidgetPropsUtils.tsx:824]: can u change the zero to a string and check if the migration works?

**Review by marks0351 — CHANGES_REQUESTED** [review]

**Comment by somangshu** [comment]
@techbhavin @marks0351 this still doesn't work of the older tables. Here is my [app](https://deploy-preview-5930--appsmith-editor-ce.netlify.app/applications/60aba4455123fa30d6c393f7/pages/60aba4455123fa30d6c393f9/edit)

**Comment by marks0351** [comment]
> @techbhavin @marks0351 this still doesn't work of the older tables. Here is my [app](https://deploy-preview-5930--appsmith-editor-ce.netlify.app/applications/60aba4455123fa30d6c393f7/pages/60aba4455123fa30d6c393f9/edit)

looks to be working for me. any chance you are using the same app as before to check this?

**Comment by somangshu** [comment]
@marks0351 its the same app as before, Should the old tables have default 0th key selected?

**Comment by marks0351** [comment]
> @marks0351 its the same app as before, Should the old tables have default 0th key selected?

by setting default selected row to 0, i thought that was what we are going for else there is no need for migration.
also if you are trying on the same app the old migration should have already run and this updated migration will not run on it, re create the app on release again and then check it on the dp if you wish to test.

**Comment by somangshu** [comment]
Thanks for the pro tip @marks0351, That worked!

**Review by somangshu — DISMISSED** [review]
Approving for the tests to run

**Review by marks0351 — DISMISSED** [review]

**Comment by somangshu** [comment]
@techbhavin can you resolve this conflict soon?

**Review by somangshu — APPROVED** [review]

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
@techbhavin this issue is fixed, Can you please merge the latest release here?

**Review by somangshu — APPROVED** [review]

**Pull request merged**